core/src/blockchain/route.rs

```diff
@@ -120,26 +120,73 @@ pub struct ImportRoute {
 }
 
 impl ImportRoute {
-    pub fn new(hash: H256, location: &BlockLocation) -> Self {
-        match location {
-            BlockLocation::CanonChain => ImportRoute {
+    pub fn new(new_block_hash: H256, best_block_changed: &BestBlockChanged) -> Self {
+        let mut omitted = Vec::new();
+        let mut enacted = Vec::new();
+        if best_block_changed.new_best_hash() == Some(*&new_block_hash) {
+            enacted.push(new_block_hash);
+        } else {
+            omitted.push(new_block_hash);
+        }
+        match best_block_changed {
+            BestBlockChanged::CanonChainAppended {
+                ..
+            } => ImportRoute {
+                retracted: vec![],
+                enacted,
+                omitted,
+            },
+            BestBlockChanged::None => ImportRoute {
+                retracted: vec![],
+                enacted,
+                omitted,
+            },
+            BestBlockChanged::BranchBecomingCanonChain {
+                tree_route,
+                ..
+            } => {
+                enacted.extend(tree_route.enacted.iter());
```

Comment:
According to `TreeRoute`'s documentation, elements of `enacted` are sorted in ascending order by their block number.
Let's put the best header at the tail to be consistent with original implementation.